docs/content/en/docs/Design/secret.md

```diff
@@ -0,0 +1,249 @@
+      %extracted = tensor.extract %arg2[%arg4] : tensor<8xi16>
+      %extracted_0 = tensor.extract %arg3[%arg4] : tensor<8xi16>
+      %2 = arith.muli %extracted, %extracted_0 : i16
+      %3 = arith.addi %arg5, %2 : i16
+      affine.yield %3 : i16
+    }
+    secret.yield %1 : i16
+  } -> !secret.secret<i16>
+  return %0 : !secret.secret<i16>
+}
+```
+
+The operands to the `generic` op are the secret data types,
+and the op contains a single region, whose block arguments are the corresponding
+cleartext data values.
+Then the region is free to perform any computation,
+and the values passed to `secret.yield` are lifted back to `secret` types.
+Note that `secret.generic` is not isolated from its enclosing scope,
+so one may refer to cleartext SSA values without adding them as generic operands
+and block arguments.
+
+Clearly `secret.generic` does not actually do anything.
+It is not decrypting data. It is merely describing the operation that one wishes
+to apply to the secret data in more familiar terms.
+It is a structural operation, primarily used to demarcate which operations
+involve secret operands and have secret results, and group them for later optimization.
+The benefit of this is that one can write optimization passes
+on types and ops that are not aware of `secret`, and they will naturally
+match on the bodies of `generic` ops.
+
+For example, here is what the above dot product computation looks like
+after applying the `-cse -canonicalize -heir-simd-vectorizer` passes,
+the implementations of which do not depend on `secret` or `generic`.
+
+```mlir
+func.func @dot_product(
+    %arg0: !secret.secret<tensor<8xi16>>,
+    %arg1: !secret.secret<tensor<8xi16>>) -> !secret.secret<i16> {
+  %c1 = arith.constant 1 : index
+  %c2 = arith.constant 2 : index
+  %c4 = arith.constant 4 : index
+  %c7 = arith.constant 7 : index
+  %0 = secret.generic ins(%arg0, %arg1 : !secret.secret<tensor<8xi16>>, !secret.secret<tensor<8xi16>>) {
+  ^bb0(%arg2: tensor<8xi16>, %arg3: tensor<8xi16>):
+    %1 = arith.muli %arg2, %arg3 : tensor<8xi16>
+    %2 = tensor_ext.rotate %1, %c4 : tensor<8xi16>, index
+    %3 = arith.addi %1, %2 : tensor<8xi16>
+    %4 = tensor_ext.rotate %3, %c2 : tensor<8xi16>, index
+    %5 = arith.addi %3, %4 : tensor<8xi16>
+    %6 = tensor_ext.rotate %5, %c1 : tensor<8xi16>, index
+    %7 = arith.addi %5, %6 : tensor<8xi16>
+    %extracted = tensor.extract %7[%c7] : tensor<8xi16>
+    secret.yield %extracted : i16
+  } -> !secret.secret<i16>
+  return %0 : !secret.secret<i16>
+}
+```
+
+The canonicalization patterns for `secret.generic` apply a variety of simplifications, such as:
+
+- Removing any unused or non-secret arguments and return values.
+- Hoisting operations in the body of a `generic` that only depend on cleartext values to the enclosing scope.
+- Removing any `generic` ops that use no secrets at all.
+
+These can be used together with the [`secret-distribute-generic`
```

Comment:
It might be worth mentioning that the reason we do split to a single op is that it ends up performing secret type prorogation to the individual ops contained in larger generics. (Removing the need to perform dataflow to figure out whether an op operand contained within a large generic block has a secret type or not)

Reply:
If you wanted, you could add a mul_plain operation as example in the snippet below

Reply:
Done.